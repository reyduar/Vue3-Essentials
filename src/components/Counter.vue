<template>
  <div>
    <h1>{{ title }} - {{ initialCount }}</h1>
    <p>{{ count }} <sup>2</sup> = {{ squareCounter }}</p>
    <p></p>
    <button @click="increment">Increment</button>
    <button @click="decrement">Decrement</button>
  </div>
</template>

<script>
export default {
  props: {
    title: String,
    initialCount: {
      type: Number,
      default: 100,
      validator: (value) => value >= 0,
    },
  },
  data() {
    return {
      count: this.initialCount,
    };
  },
  methods: {
    increment() {
      this.count++;
    },
    decrement() {
      this.count--;
    },
    getSquared() {
      return this.count * this.count;
    },
  },
  computed: {
    squareCounter() {
      return this.count * this.count;
    },
    customTitle() {
      return this.title ? this.title : "Counter";
    },
  },
};
</script>

<style>
button {
  margin: 0 10px;
  padding: 10px 20px;
  border-radius: 5px;
  background-color: #4caf50;
  color: white;
  border: none;
  transition: 0.3s ease-in-out;
}

button:hover {
  background-color: #45a049;
  transition: 0.3s ease-in-out;
}
</style>
