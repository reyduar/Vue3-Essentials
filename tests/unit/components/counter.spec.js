import Counter from "@/components/Counter.vue";
import { shallowMount } from "@vue/test-utils";

describe("Counter Component test", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallowMount(Counter);
  });

  test("should match with the snapshot", () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  test("h1 should have default value", () => {
    const h1 = wrapper.find("h1");
    expect(h1.exists()).toBeTruthy();
    expect(h1.text()).toBe("Counter");
  });

  test("value should start with 100", () => {
    const value = wrapper.find('[data-testid="initialCount"]');
    expect(value.text()).toBe("100");
  });

  test("should increment the value", async () => {
    const increaseBtn = wrapper.find("button");

    await increaseBtn.trigger("click");
    const value = wrapper.find('[data-testid="counter"]');

    expect(value.text()).toBe("101");
  });

  test("should decrement the value", async () => {
    const decreaseBtn = wrapper.findAll("button").at(1);

    await decreaseBtn.trigger("click");
    const value = wrapper.find('[data-testid="counter"]');
    console.log(value.text());

    expect(value.text()).toBe("99");
  });

  test("should increment and decrement counter", async () => {
    const [increaseBtn, decreaseBtn] = wrapper.findAll("button");

    await increaseBtn.trigger("click");
    await increaseBtn.trigger("click");
    await increaseBtn.trigger("click");
    await decreaseBtn.trigger("click");
    await decreaseBtn.trigger("click");
    const value = wrapper.find('[data-testid="counter"]');
    console.log(value.text());

    expect(value.text()).toBe("101");
  });
});
