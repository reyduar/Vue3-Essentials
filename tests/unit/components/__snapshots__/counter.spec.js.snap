// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Component test should match with the snapshot 1`] = `
<div>
  <h1>Counter</h1>
  <h2 data-testid="initialCount">100</h2>
  <p><span data-testid="counter">100</span><sup>2</sup> = 10000</p>
  <p></p><button>Increment</button><button>Decrement</button>
</div>
`;
